File: app/src/main/java/com/example/recipeapp/persistence/RecipeDAO.java
package com.example.recipeapp.persistence;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import com.example.recipeapp.models.Ingredient;
import com.example.recipeapp.models.Recipe;
import com.example.recipeapp.models.RecipeWithIngredients;

import java.util.List;

import static androidx.room.OnConflictStrategy.REPLACE;

@Dao
public abstract class RecipeDAO {

    @Query("SELECT * FROM recipe_table")
    public abstract List<Recipe> getAll();

//    @Insert(onConflict = OnConflictStrategy.IGNORE)
//    public abstract long insert(Recipe recipe);

    @Delete
    public abstract void delete(Recipe recipe);

    @Update
    public abstract void update(Recipe recipe);

    @Update
    public abstract void update(Ingredient ingredient);

    @Query("DELETE FROM recipe_table")
    public abstract void DeleteAll();

    /*
    method to insert a recipe with the corresponding ingredients
    source: https://github.com/relativizt/android-room-one-to-many-auto-pk/blob/master/app/src/main/java/com/arc/roomonetomany/datasource/EmployeeDao.java
     */
    @Transaction
    @Query("SELECT * FROM recipe_table")
    public abstract LiveData<List<RecipeWithIngredients>> getRecipesWithIngredients();

    @Transaction
    public void insert(RecipeWithIngredients recipeWithIngredients) {

        // Save rowId of inserted Recipe as recipeId
        final long recipeId = insert(recipeWithIngredients.getRecipe());

        // Set recipeId for all related ingredientEntities
        for (Ingredient ingredient : recipeWithIngredients.getIngredientList()) {
            ingredient.setBelongsToRecipeID(recipeId);
            insert(ingredient);
        }

    }

    @Transaction
    public void update(RecipeWithIngredients recipeWithIngredients){
        update(recipeWithIngredients.getRecipe());
        for(int i = 0; i < recipeWithIngredients.getIngredientList().size(); i ++){
            Ingredient ingredient = recipeWithIngredients.getIngredientList().get(i);
            ingredient.setBelongsToRecipeID(recipeWithIngredients.getRecipe().getRecipeId());
            long id = insert(ingredient);
            if (id == -1) {
                update(ingredient);
            }
        }

    }

    // If the @Insert method receives only 1 parameter, it can return a long,
    // which is the new rowId for the inserted item.
    // https://developer.android.com/training/data-storage/room/accessing-data
    @Insert(onConflict = REPLACE)
    public abstract long insert(Recipe recipe);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    public abstract long insert(Ingredient ingredient);

}

File: app/src/main/java/com/example/recipeapp/persistence/RecipeDAO.kt
package com.example.recipeapp.persistence

import androidx.lifecycle.LiveData
import androidx.room.*
import com.example.recipeapp.models.Ingredient
import com.example.recipeapp.models.Recipe
import com.example.recipeapp.models.RecipeWithIngredients

@Dao
abstract class RecipeDAO {
    @get:Query("SELECT * FROM recipe_table")
    abstract val all: List<Recipe?>?

    //    @Insert(onConflict = OnConflictStrategy.IGNORE)
    //    public abstract long insert(Recipe recipe);
    @Delete
    abstract fun delete(recipe: Recipe?)

    @Update
    abstract fun update(recipe: Recipe?)

    @Update
    abstract fun update(ingredient: Ingredient?)

    @Query("DELETE FROM recipe_table")
    abstract fun DeleteAll()

    /*
    method to insert a recipe with the corresponding ingredients
    source: https://github.com/relativizt/android-room-one-to-many-auto-pk/blob/master/app/src/main/java/com/arc/roomonetomany/datasource/EmployeeDao.java
     */
    @get:Query("SELECT * FROM recipe_table")
    @get:Transaction
    abstract val recipesWithIngredients: LiveData<List<RecipeWithIngredients?>?>?

    @Transaction
    open fun insert(recipeWithIngredients: RecipeWithIngredients) {

        // Save rowId of inserted Recipe as recipeId
        val recipeId = insert(recipeWithIngredients.recipe)

        // Set recipeId for all related ingredientEntities
        for (ingredient in recipeWithIngredients.ingredientList!!) {
            ingredient!!.belongsToRecipeID = recipeId
            insert(ingredient)
        }
    }

    @Transaction
    open fun update(recipeWithIngredients: RecipeWithIngredients) {
        update(recipeWithIngredients.recipe)
        for (i in recipeWithIngredients.ingredientList!!.indices) {
            val ingredient = recipeWithIngredients.ingredientList!![i]!!
            ingredient.belongsToRecipeID = recipeWithIngredients.recipe!!.recipeId
            val id = insert(ingredient)
            if (id == -1L) {
                update(ingredient)
            }
        }
    }

    // If the @Insert method receives only 1 parameter, it can return a long,
    // which is the new rowId for the inserted item.
    // https://developer.android.com/training/data-storage/room/accessing-data
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insert(recipe: Recipe?): Long

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    abstract fun insert(ingredient: Ingredient?): Long
}